### PR TITLE
Fix crashing benchmarks

### PR DESCRIPTION
**Pull request by TheoBerlin**
## Purpose
- Benchmarks no longer crash
- The characters in the benchmarks no longer run out of ammo after emptying one magazine

## Changes
- Crash fixes:
  - Added `WeaponNetworkUID` to `CreatePlayerDesc` in `BenchmarkState`
  - Fixed shared pointers being deleted twice in `~PaintMaskRenderer`
- `BenchmarkSystem` takes care of weapon-firing logic (very simplified version of the logic in `WeaponSystem`)

### CODE REVIEW

File: CrazyCanvas/Source/ECS/Systems/Player/BenchmarkSystem.cpp [hunk 1]
```diff
@@ -60,22 +62,20 @@ void BenchmarkSystem::Tick(LambdaEngine::Timestamp deltaTime)
 	const float32 dt = (float32)deltaTime.AsSeconds();
 
 	ECSCore* pECS = ECSCore::GetInstance();
-	ComponentArray<WeaponComponent>* pWeaponComponents		= pECS->GetComponentArray<WeaponComponent>();
-	ComponentArray<RotationComponent>* pRotationComponents	= pECS->GetComponentArray<RotationComponent>();
+
+	ComponentArray<WeaponComponent>* pWeaponComponents				= pECS->GetComponentArray<WeaponComponent>();
+	const ComponentArray<PositionComponent>* pPositionComponents	= pECS->GetComponentArray<PositionComponent>();
+	ComponentArray<RotationComponent>* pRotationComponents			= pECS->GetComponentArray<RotationComponent>();
```
Comment (frozenbanana) [inline CrazyCanvas/Source/ECS/Systems/Player/BenchmarkSystem.cpp:69]: Should be `const ComponentArray<..>* pBlablaComponents = pECS->GetConstComponentArray<..>();`?

File: CrazyCanvas/Source/States/BenchmarkState.cpp [hunk 2]
```diff
@@ -305,12 +314,13 @@ bool BenchmarkState::OnPacketReceived(const LambdaEngine::NetworkSegmentReceived
...
 
-			for (uint32 playerNr = 0; playerNr < 9; playerNr++)
+			for (uint32 playerNr = 0; playerNr < playerCount; playerNr++)
 			{
 				// Create a 3x3 grid of players in the XZ plane
 				createPlayerDesc.Position.x			= -3.0f + 3.0f * (playerNr % 3);
 				createPlayerDesc.Position.z			= -3.0f + 3.0f * (playerNr / 3);
-				createPlayerDesc.PlayerNetworkUID	+= (int32)playerNr;
+				createPlayerDesc.PlayerNetworkUID	+= 2;
+				createPlayerDesc.WeaponNetworkUID	+= 2;
```
Comment (frozenbanana) [inline CrazyCanvas/Source/States/BenchmarkState.cpp:322]: Why do we need to increment by 2?

File: CrazyCanvas/Source/ECS/Systems/Player/BenchmarkSystem.cpp (hunk 1 above)
Comment (TheoBerlin) [reply CrazyCanvas/Source/ECS/Systems/Player/BenchmarkSystem.cpp:69]: No, since rotation components are written to.

**Review by Ceanze — APPROVED** [review]
Yes it is a very a good 👍

File: CrazyCanvas/Source/States/BenchmarkState.cpp (hunk 2 above)
Comment (TheoBerlin) [reply CrazyCanvas/Source/States/BenchmarkState.cpp:322]: `PlayerNetworkUID` starts at 0, `WeaponNetworkUID` at 1. If both are incremented by 2, players and weapons will never be the same.

**Review by frozenbanana — CHANGES_REQUESTED** [review]
Looks good overall. Left some comments and also I don't understand why lines were removed in `PaintMaskRenderer.cpp` is it related to the bench mark fix?

**Comment by TheoBerlin** [comment]
> Looks good overall. Left some comments and also I don't understand why lines were removed in `PaintMaskRenderer.cpp` is it related to the bench mark fix?

Shared pointers were being deleted twice in `~PaintMaskRenderer`

**Comment by Ceanze** [comment]
> Looks good overall. Left some comments and also I don't understand why lines were removed in `PaintMaskRenderer.cpp` is it related to the bench mark fix?

When the class gets out of scope, the array that was cleared and its content will be automatically cleared. And since this is an array that contains TSharedRef its resources will also be removed. If the lines would still be there it would lower the ref count on TSharedRef too many times causing error.

**Review by frozenbanana — APPROVED** [review]
👍🏾

**Pull request merged**